### PR TITLE
Add users to media folders

### PR DESCRIPTION
**Pull request by wiltso**
- Added m2m relations between the user model and the media folder
	1. If the folder is the users root folder
- Added boolean field to media folder if it's show to everyone in the media browser
- Added so users who are in there root folder or a sub folder of their root folder can perform any action
- Added permissions so none admin users can perform different actions if they have permissions for it

refs MYYR-26

### CODE REVIEW

File: shuup/admin/settings.py [hunk 1]
```diff
@@ -117,3 +117,6 @@
 #: If this is False, only a allow simple text field and sanitize all HTML from it.
 #:
 SHUUP_ADMIN_ALLOW_HTML_IN_VENDOR_DESCRIPTION = True
+
+#: The method used to get the media folder from the folder.
+SHUUP_GET_MEDIA_FOLDER_FROM_FOLDER = "shuup.admin.utils.media_folder_from_folder"
```
Comment (tulimaki) [inline shuup/admin/settings.py:122]: Can you explain a little bit more in depth what is this for?

File: shuup/admin/modules/media/__init__.py [hunk 2]
```diff
@@ -27,13 +27,30 @@ def get_urls(self):
...
+            ),
             admin_url(
                 "^media/upload/$",
                 "shuup.admin.modules.media.views.media_upload",
                 name="media.upload"
             )
         ]
 
+    def get_extra_permissions(self):
+        permissions = super().get_extra_permissions()
```
Comment (tulimaki) [inline shuup/admin/modules/media/__init__.py:43]: What is this extra permissions for?

File: shuup/front/views/upload.py [hunk 3]
```diff
@@ -12,12 +12,11 @@
 from django.core.exceptions import ValidationError
 from django.http.response import HttpResponseForbidden, JsonResponse
 from django.utils.translation import ugettext as _
-from filer.models import Folder
 
+from shuup.admin.modules.media.views import get_or_create_folder
```
Comment (tulimaki) [inline shuup/front/views/upload.py:16]: Shuup front shouldn't know anything about admin module I think.

File: shuup/core/models/_media.py [hunk 4]
```diff
@@ -30,7 +31,24 @@ class MediaFolder(models.Model):
     folder = FilerFolderField(related_name="media_folder", verbose_name=_("folder"), on_delete=models.CASCADE)
     shops = models.ManyToManyField(
         "shuup.Shop", related_name="media_folders", verbose_name=_('shops'),
-        help_text=_("Select which shops you would like the folder to be visible in."))
+        help_text=_("Select which shops you would like the folder to be visible in.")
+    )
+
+    user_access = models.ManyToManyField(
```
Comment (tulimaki) [inline shuup/core/models/_media.py:37]: So what happens if vendor/supplier are linked to multiple users?

File: shuup/admin/modules/media/views.py [hunk 5]
```diff
@@ -193,51 +241,69 @@ def handle_get_folder(self, data):
...
+        folder = _get_folder_query(shop, self.user).get(pk=data["id"])
+        if not _is_folder_shared(folder) and (
+            subfolder_of_users_root(self.user, folder) or
+            has_permission(self.user, "media.delete-folder")
+        ):
+            new_selected_folder_id = folder.parent_id
+            message = delete_folder(folder)
+            return JsonResponse({"success": True, "message": message, "newFolderId": new_selected_folder_id})
+
+        message = _("Can't delete a this folder.")
```
Comment (tulimaki) [inline shuup/admin/modules/media/views.py:281]: Isn't this just "Can't delete this folder."?

File: shuup/admin/modules/media/views.py [hunk 6]
```diff
@@ -141,37 +161,65 @@ def post(self, request, *args, **kwargs):
...
         if parent_id > 0:
-            parent = _get_folder_query(shop).get(pk=parent_id)
+            parent = _get_folder_query(shop, self.user).get(pk=parent_id)
         else:
             parent = None
 
-        name = data["name"]
+        if parent and subfolder_of_users_root(self.user, parent):
+            return self._create_folder(data["name"], parent, shop)
+        elif has_permission(self.user, "media.create-folder"):
```
Comment (tulimaki) [inline shuup/admin/modules/media/views.py:194]: Doesn't this mean that I can always add folders anywhere if I have "create-folder" perissions.

File: shuup/front/views/upload.py (hunk 3 above)
Comment (wiltso) [reply shuup/front/views/upload.py:16]: But can front know about `shuup.utils` as it already dose?

File: shuup/admin/modules/media/views.py [hunk 7]
```diff
@@ -43,12 +53,17 @@ def _is_folder_shared(folder):
...
 
 
-def _get_folder_query_filter(shop):
-    return Q(Q(media_folder__isnull=True) | Q(media_folder__shops__isnull=True) | Q(media_folder__shops=shop))
+def _get_folder_query_filter(shop, user=None):
+    query = Q(
+        Q(Q(media_folder__isnull=True) | Q(media_folder__shops__isnull=True) | Q(media_folder__shops=shop))
+    )
+    if user and not user.is_superuser:
+        query &= Q(media_folder__user_access=user)
```
Comment (tulimaki) [inline shuup/admin/modules/media/views.py:61]: So I need to add per folder links for each shop staff member?

File: shuup/core/models/_media.py (hunk 4 above)
Comment (wiltso) [reply shuup/core/models/_media.py:37]: Then all the vendors users will have to be added here.

File: shuup/admin/modules/media/views.py (hunk 6 above)
Comment (wiltso) [reply shuup/admin/modules/media/views.py:194]: Yes

Comment (wiltso) [reply shuup/admin/modules/media/views.py:193]: But if you don't have that permission you will only be able to create folders that are in the sub folder tree of your root folder

**Pull request closed without merging**